File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/quests.rb
# frozen_string_literal: true

require 'json'
require_relative '../errors'
require_relative '../database'
require_relative 'campaigns'

module Handlers
  # Campaign quest tracking: create quests with milestones, mark milestones
  # complete, and summarize quest status counts per campaign.
  module Quests
    STATUSES = %w[active completed blocked].freeze

    module_function

    def find_quest(campaign_id, quest_id)
      row = Database.query(<<~SQL).first
        SELECT id, campaign_id, title, status, milestones_json, completed_json
        FROM campaign_quests
        WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(quest_id)};
      SQL
      raise HttpError.new(404, 'unknown quest id') unless row

      row
    end

    def create_quest(campaign_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)

      id = body['id']
      title = body['title']
      status = body['status']
      milestones = body['milestones']

      raise HttpError.new(400, 'id must be a string') unless id.is_a?(String) && !id.empty?
      raise HttpError.new(400, 'title must be a string') unless title.is_a?(String) && !title.empty?
      raise HttpError.new(400, 'status must be one of active, completed, blocked') unless STATUSES.include?(status)
      raise HttpError.new(400, 'milestones must be an array of strings') unless milestones.is_a?(Array) && milestones.all? { |m| m.is_a?(String) }
      raise HttpError.new(409, 'id already exists') unless Database.query("SELECT id FROM campaign_quests WHERE id = #{Database.escape(id)};").empty?

      Database.exec(<<~SQL)
        INSERT INTO campaign_quests (id, campaign_id, title, status, milestones_json, completed_json)
        VALUES (
          #{Database.escape(id)},
          #{Database.escape(campaign_id)},
          #{Database.escape(title)},
          #{Database.escape(status)},
          #{Database.escape(JSON.generate(milestones))},
          #{Database.escape(JSON.generate([]))}
        );
      SQL

      [201, {
        id: id,
        title: title,
        status: status,
        milestones_total: milestones.size,
        milestones_done: 0
      }]
    end

    def update_progress(campaign_id, quest_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)
      quest = find_quest(campaign_id, quest_id)

      completed_arg = body['completed']
      raise HttpError.new(400, 'completed must be an array of strings') unless completed_arg.is_a?(Array) && completed_arg.all? { |m| m.is_a?(String) }

      milestones = JSON.parse(quest['milestones_json'])
      completed = JSON.parse(quest['completed_json'])

      invalid = completed_arg - milestones
      raise HttpError.new(400, "unknown milestone(s): #{invalid.join(', ')}") unless invalid.empty?

      completed = (completed + completed_arg).uniq

      Database.exec(<<~SQL)
        UPDATE campaign_quests
        SET completed_json = #{Database.escape(JSON.generate(completed))}
        WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(quest_id)};
      SQL

      [200, {
        id: quest_id,
        status: quest['status'],
        milestones_total: milestones.size,
        milestones_done: completed.size
      }]
    end

    def summary(campaign_id, _body)
      Handlers::Campaigns.find_campaign(campaign_id)

      rows = Database.query("SELECT status FROM campaign_quests WHERE campaign_id = #{Database.escape(campaign_id)};")

      counts = { 'active' => 0, 'completed' => 0, 'blocked' => 0 }
      rows.each { |row| counts[row['status']] += 1 if counts.key?(row['status']) }

      [200, {
        campaign_id: campaign_id,
        active: counts['active'],
        completed: counts['completed'],
        blocked: counts['blocked']
      }]
    end
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081336Z_claude_claude-opus-5_go-stdlib/quests.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"
)

// Campaign quest tracking: quests hang off a campaign, carry an ordered
// milestone checklist, and record which milestones are done.
//
// A quest owns two parallel slices: Milestones (the checklist text, in the
// order supplied) and Done (one flag per milestone). Keeping them parallel —
// rather than a set of completed titles — means duplicate milestone text stays
// countable, since each entry is ticked off independently.
//
// Like the rest of the campaign state, quests live in memory under
// campaigns.mu and are mirrored to SQLite by flush().

// Quest lifecycle statuses. A quest may be created in any of them, and reaching
// the last milestone promotes an active quest to completed.
const (
	questActive    = "active"
	questCompleted = "completed"
	questBlocked   = "blocked"
)

// questStatuses is the accepted set, listed in summary-response order.
var questStatuses = []string{questActive, questCompleted, questBlocked}

func validQuestStatus(s string) bool {
	for _, known := range questStatuses {
		if s == known {
			return true
		}
	}
	return false
}

type quest struct {
	ID         string
	Title      string
	Status     string
	Milestones []string
	Done       []bool
}

// doneCount reports how many milestones are ticked off.
func (q *quest) doneCount() int {
	n := 0
	for _, d := range q.Done {
		if d {
			n++
		}
	}
	return n
}

// complete marks the first not-yet-done milestone whose text equals name,
// reporting whether such a milestone exists at all. Re-completing an already
// finished milestone is a no-op success, so replayed progress calls are
// idempotent.
func (q *quest) complete(name string) bool {
	found := false
	for i, m := range q.Milestones {
		if m != name {
			continue
		}
		found = true
		if !q.Done[i] {
			q.Done[i] = true
			return true
		}
	}
	return found
}

// settle promotes an active quest to completed once every milestone is done. A
// blocked or already-completed quest keeps its status, and a quest with no
// milestones is never auto-completed.
func (q *quest) settle() {
	if q.Status != questActive || len(q.Milestones) == 0 {
		return
	}
	if q.doneCount() == len(q.Milestones) {
		q.Status = questCompleted
	}
}

// find returns the campaign's quest with the given id. Callers must hold
// campaigns.mu.
func findQuest(c *campaign, id string) *quest {
	for _, q := range c.Quests {
		if q.ID == id {
			return q
		}
	}
	return nil
}

// ---------- request / response payloads ----------

type questRequest struct {
	ID         *string   `json:"id"`
	Title      *string   `json:"title"`
	Status     *string   `json:"status"`
	Milestones *[]string `json:"milestones"`
}

type questResponse struct {
	ID              string `json:"id"`
	Title           string `json:"title"`
	Status          string `json:"status"`
	MilestonesTotal int    `json:"milestones_total"`
	MilestonesDone  int    `json:"milestones_done"`
}

type questProgressRequest struct {
	Completed *[]string `json:"completed"`
	Status    *string   `json:"status"`
}

type questProgressResponse struct {
	ID              string `json:"id"`
	Status          string `json:"status"`
	MilestonesTotal int    `json:"milestones_total"`
	MilestonesDone  int    `json:"milestones_done"`
}

type questSummaryResponse struct {
	CampaignID string `json:"campaign_id"`
	Active     int    `json:"active"`
	Completed  int    `json:"completed"`
	Blocked    int    `json:"blocked"`
}

// ---------- POST /v1/campaigns/{id}/quests ----------

func handleCreateQuest(w http.ResponseWriter, r *http.Request) {
	var req questRequest
	if err := decodeBody(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	id, ok := requiredString(req.ID)
	if !ok {
		writeError(w, http.StatusBadRequest, "id is required")
		return
	}
	title, ok := requiredString(req.Title)
	if !ok {
		writeError(w, http.StatusBadRequest, "title is required")
		return
	}
	status, ok := requiredString(req.Status)
	if !ok {
		writeError(w, http.StatusBadRequest, "status is required")
		return
	}
	status = strings.ToLower(status)
	if !validQuestStatus(status) {
		writeError(w, http.StatusBadRequest, "status must be active, completed, or blocked")
		return
	}
	// Milestones are optional; an omitted list simply means a quest with
	// nothing to tick off yet. A present list must hold non-blank strings.
	milestones := []string{}
	if req.Milestones != nil {
		for _, m := range *req.Milestones {
			m = strings.TrimSpace(m)
			if m == "" {
				writeError(w, http.StatusBadRequest, "milestones must be non-empty strings")
				return
			}
			milestones = append(milestones, m)
		}
	}

	campaigns.mu.Lock()
	c, found := campaigns.campaigns[r.PathValue("id")]
	if !found {
		campaigns.mu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	if findQuest(c, id) != nil {
		campaigns.mu.Unlock()
		writeError(w, http.StatusConflict, "quest id already exists")
		return
	}
	q := &quest{
		ID:         id,
		Title:      title,
		Status:     status,
		Milestones: milestones,
		Done:       make([]bool, len(milestones)),
	}
	c.Quests = append(c.Quests, q)
	resp := questResponse{
		ID:              q.ID,
		Title:           q.Title,
		Status:          q.Status,
		MilestonesTotal: len(q.Milestones),
		MilestonesDone:  q.doneCount(),
	}
	campaigns.mu.Unlock()
	flush()

	writeJSON(w, http.StatusCreated, resp)
}

// ---------- POST /v1/campaigns/{id}/quests/{quest_id}/progress ----------

func handleQuestProgress(w http.ResponseWriter, r *http.Request) {
	var req questProgressRequest
	if err := decodeBody(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	if req.Completed == nil {
		writeError(w, http.StatusBadRequest, "completed is required")
		return
	}
	// An explicit status is optional and lets a quest be blocked or reopened
	// without touching its checklist.
	status := ""
	if req.Status != nil {
		s, ok := requiredString(req.Status)
		if !ok {
			writeError(w, http.StatusBadRequest, "status must not be blank")
			return
		}
		status = strings.ToLower(s)
		if !validQuestStatus(status) {
			writeError(w, http.StatusBadRequest, "status must be active, completed, or blocked")
			return
		}
	}

	campaigns.mu.Lock()
	c, found := campaigns.campaigns[r.PathValue("id")]
	if !found {
		campaigns.mu.Unlock()
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	q := findQuest(c, r.PathValue("quest_id"))
	if q == nil {
		campaigns.mu.Unlock()
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}
	// Validate the whole batch before applying any of it, so a bad name cannot
	// leave the quest half-updated.
	names := make([]string, 0, len(*req.Completed))
	for _, name := range *req.Completed {
		name = strings.TrimSpace(name)
		if name == "" {
			campaigns.mu.Unlock()
			writeError(w, http.StatusBadRequest, "completed must be non-empty strings")
			return
		}
		known := false
		for _, m := range q.Milestones {
			if m == name {
				known = true
				break
			}
		}
		if !known {
			campaigns.mu.Unlock()
			writeError(w, http.StatusBadRequest, "unknown milestone")
			return
		}
		names = append(names, name)
	}
	for _, name := range names {
		q.complete(name)
	}
	if status != "" {
		q.Status = status
	}
	q.settle()
	resp := questProgressResponse{
		ID:              q.ID,
		Status:          q.Status,
		MilestonesTotal: len(q.Milestones),
		MilestonesDone:  q.doneCount(),
	}
	campaigns.mu.Unlock()
	flush()

	writeJSON(w, http.StatusOK, resp)
}

// ---------- GET /v1/campaigns/{id}/quests/summary ----------

func handleQuestSummary(w http.ResponseWriter, r *http.Request) {
	campaigns.mu.Lock()
	defer campaigns.mu.Unlock()
	c, ok := campaigns.campaigns[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	resp := questSummaryResponse{CampaignID: c.ID}
	for _, q := range c.Quests {
		switch q.Status {
		case questActive:
			resp.Active++
		case questCompleted:
			resp.Completed++
		case questBlocked:
			resp.Blocked++
		}
	}
	writeJSON(w, http.StatusOK, resp)
}

// ---------- persistence helpers ----------

// questRows renders one campaign's quests as storage rows. Milestone text and
// the done flags are JSON-encoded because the writer only handles flat columns.
// Callers must hold campaigns.mu.
func questRows(c *campaign) [][]any {
	out := make([][]any, 0, len(c.Quests))
	for i, q := range c.Quests {
		milestones, _ := json.Marshal(q.Milestones)
		done, _ := json.Marshal(q.Done)
		out = append(out, []any{
			c.ID, q.ID, q.Title, q.Status, string(milestones), string(done), int64(i),
		})
	}
	return out
}

// questFromRow rebuilds a quest from a storage row, returning the owning
// campaign id. A row whose flag list disagrees with its milestone list is
// rejected, since the two must stay parallel.
func questFromRow(row []any) (campaignID string, q *quest, ok bool) {
	if len(row) < 6 {
		return "", nil, false
	}
	campaignID, _ = row[0].(string)
	id, _ := row[1].(string)
	title, _ := row[2].(string)
	status, _ := row[3].(string)
	milestonesJSON, _ := row[4].(string)
	doneJSON, _ := row[5].(string)
	if campaignID == "" || id == "" || !validQuestStatus(status) {
		return "", nil, false
	}
	milestones := []string{}
	_ = json.Unmarshal([]byte(milestonesJSON), &milestones)
	done := []bool{}
	_ = json.Unmarshal([]byte(doneJSON), &done)
	if len(done) != len(milestones) {
		return "", nil, false
	}
	return campaignID, &quest{
		ID:         id,
		Title:      title,
		Status:     status,
		Milestones: milestones,
		Done:       done,
	}, true
}
